data/transactions/logic/eval.go

```diff
@@ -612,26 +612,226 @@ func (cx *EvalContext) RunMode() RunMode {
+	StackBigInt = NewStackType(avmBytes, bound(0, maxByteMathSize), "bigint")
+	// StackMethodSelector represents a bytestring that should be treated like a method selector
+	StackMethodSelector = NewStackType(avmBytes, static(4), "method")
+	// StackStorageKey represents a bytestring that can be used as a key to some storage (global/local/box)
+	StackStorageKey = NewStackType(avmBytes, bound(0, 64), "key")
+
+	// AllStackTypes is a list of all the stack types we recognize
+	// so that we can iterate over them in doc prep
+	AllStackTypes = []StackType{
+		StackUint64,
+		StackBytes,
+		StackAny,
+		StackNone,
+		StackBoolean,
+		StackHash,
+		StackAddress,
+		StackBigInt,
+		StackMethodSelector,
+		StackStorageKey,
+	}
 )
 
+func bound(min, max uint64) [2]uint64 {
+	return [2]uint64{min, max}
+}
+
+func static(size uint64) [2]uint64 {
+	return bound(size, size)
+}
+
+// StackType describes the type of a value on the operand stack
+type StackType struct {
+	Name        string
+	AVMType     avmType
+	LengthBound [2]uint64
+	ValueBound  [2]uint64
+}
+
+// NewStackType Initializes a new StackType with fields passed
+func NewStackType(at avmType, bounds [2]uint64, stname ...string) StackType {
+	name := at.String()
+	if len(stname) > 0 {
+		name = stname[0]
+	}
+
+	st := StackType{Name: name, AVMType: at}
+
+	switch at {
+	case avmBytes:
+		st.LengthBound = bounds
+	case avmUint64:
+		st.ValueBound = bounds
+	}
+
+	return st
+}
+
+func (st StackType) narrowed(bounds [2]uint64) StackType {
+	// It's static, set the name to show
+	// the static value
+	if bounds[0] == bounds[1] {
+		switch st.AVMType {
+		case avmBytes:
+			return NewStackType(st.AVMType, bounds, fmt.Sprintf("[%d]byte", bounds[0]))
+		case avmUint64:
+			return NewStackType(st.AVMType, bounds, fmt.Sprintf("%d", bounds[0]))
+		}
+	}
+	return NewStackType(st.AVMType, bounds)
+}
+
+// AssignableTo returns a bool indicating whether the receiver can be
+// assigned to some other type that is expected by the next operation
+func (st StackType) AssignableTo(other StackType) bool {
+	if st.AVMType == avmNone || other.AVMType == avmNone {
+		return false
+	}
+
+	if st.AVMType == avmAny || other.AVMType == avmAny {
+		return true
+	}
+
+	// By now, both are either uint or bytes
+	// and must match
+	if st.AVMType != other.AVMType {
+		return false
+	}
+
+	// Same type now
+
+	// Check if our constraints will be satisfied by
+	// the other type
+	switch st.AVMType {
+	case avmBytes:
+		smin, smax := st.LengthBound[0], st.LengthBound[1]
+		omin, omax := other.LengthBound[0], other.LengthBound[1]
+
+		// yes definitely
+		// [32,32] => [0..4k]
+		// [32,32] => [32,32]
+
+		// yes, maybe determined at runtime
+		// [0..4k] => [32,32]
+
+		// no, cant fit
+		// [64,64] => [32,32]
+		// no, makes no sense
+		// [32,32] =>  [64,64]
+
+		// we only have 0-N and [N,N] (static) and only
+		// those that are both not static and have different lengths
+		// can be assigned
```

Comment:
I'd rather we do the real checks (that there exists at least some overlap) than make the simplifying assumption about when kinds of constraints we actually use.